contracts/instruments/RibbonThetaVaultYearn.sol

```diff
@@ -0,0 +1,815 @@
+            "Wrong collateralAsset"
+        );
+
+        // we just assume all options use USDC as the strike
+        require(otoken.strikeAsset() == USDC, "strikeAsset != USDC");
+
+        uint256 readyAt = block.timestamp.add(delay);
+        require(
+            otoken.expiryTimestamp() >= readyAt,
+            "Option expiry cannot be before delay"
+        );
+
+        nextOption = option;
+        nextOptionReadyAt = readyAt;
+    }
+
+    /**
+     * @notice Closes the existing short position for the vault.
+     */
+    function _closeShort() private {
+        address oldOption = currentOption;
+        currentOption = address(0);
+        lockedAmount = 0;
+
+        if (oldOption != address(0)) {
+            OtokenInterface otoken = OtokenInterface(oldOption);
+            require(
+                block.timestamp > otoken.expiryTimestamp(),
+                "Cannot close short before expiry"
+            );
+            uint256 withdrawAmount = adapter.delegateCloseShort();
+            emit CloseShort(oldOption, withdrawAmount, msg.sender);
+        }
+    }
+
+    /**
+     * @notice Rolls the vault's funds into a new short position.
+     */
+    function rollToNextOption() external onlyManager nonReentrant {
+        require(
+            block.timestamp >= nextOptionReadyAt,
+            "Cannot roll before delay"
+        );
+
+        address newOption = nextOption;
+        require(newOption != address(0), "No found option");
+
+        currentOption = newOption;
+        nextOption = address(0);
+
+        uint256 amountToWrap = IERC20(asset).balanceOf(address(this));
+        // prevent spending old and new allowance through tricky tx ordering
+        IERC20(asset).safeApprove(address(collateralToken), 0);
+        IERC20(asset).safeApprove(address(collateralToken), amountToWrap);
+        collateralToken.deposit(amountToWrap, address(this));
+
+        uint256 currentBalance = assetBalance();
+        (uint256 queuedWithdrawAmount, , ) =
+            _withdrawAmountWithShares(queuedWithdrawShares, currentBalance);
+        uint256 freeBalance = currentBalance.sub(queuedWithdrawAmount);
+        uint256 shortAmount = wmul(freeBalance, lockedRatio);
+        lockedAmount = shortAmount;
+
+        OtokenInterface otoken = OtokenInterface(newOption);
+
+        ProtocolAdapterTypes.OptionTerms memory optionTerms =
+            ProtocolAdapterTypes.OptionTerms(
+                otoken.underlyingAsset(),
+                USDC,
+                otoken.collateralAsset(),
+                otoken.expiryTimestamp(),
+                otoken.strikePrice().mul(10**10), // scale back to 10**18
+                isPut
+                    ? ProtocolAdapterTypes.OptionType.Put
+                    : ProtocolAdapterTypes.OptionType.Call, // isPut
+                address(0)
+            );
+
+        uint256 shortAmountInYieldToken =
+            wdiv(shortAmount, collateralToken.pricePerShare());
```

Comment:
Why do we need to do wdiv here? We have already gotten the `shortAmount` which is denominated in the yield token already

Reply:
I see. It's a bit of a confusing read to me because `assetBalance` also includes the [yvWETH token balance](https://github.com/ribbon-finance/structured-products/pull/139/files/2f46ede94d794c588931a60f5d0b1ac6596c8cd7#diff-9bf745fa6d09464c08efea295ca698544ee7cde8390915fcf895905fec4128b0R793-R798).

Maybe a clearer way to write it is to assume that ALL `asset` tokens have been wrapped, and just use that number? Like so:

```
wmul(IERC20(collateraToken).balanceOf(address(this)), 0.9 ether)
```

Reply:
this doesn't account for queuedWithdrawAmount > 0 which is denominated in eth terms so it will get messy either way. I do see why its confusing. I can add a comment to clear it up a bit.

Reply:
UPDATE: inverted code bit that solves this and new getter